ergm.py

```diff
@@ -237,96 +274,121 @@ def fit(self, observed_network,
         print("optimization started")
 
         self.optimization_start_time = time.time()
+        num_of_features = self._network_statistics.get_num_of_features(self._n_nodes)
+        
+        if convergence_criterion == "hotelling":
+            hotelling_critical_value = f.ppf(1-hotelling_confidence, num_of_features, self.sample_size - num_of_features) # F(p, n-p) TODO - doc this better
 
-        grads = np.zeros((opt_steps, self._network_statistics.get_num_of_features(self._n_nodes)))
-        true_grads = np.zeros((opt_steps, self._network_statistics.get_num_of_features(self._n_nodes)))
+        grads = np.zeros((opt_steps, num_of_features))
+        hotelling_statistics = []
+        
+        prev_cov_matrix = np.ones((num_of_features, num_of_features))
 
         for i in range(opt_steps):
             if ((i + 1) % steps_for_decay) == 0:
                 lr *= (1 - lr_decay_pct)
 
-                if self.n_networks_for_grad_estimation < max_nets_for_sample:
-                    self.n_networks_for_grad_estimation *= (1 + sample_pct_growth)
-                    self.n_networks_for_grad_estimation = np.min(
-                        [int(self.n_networks_for_grad_estimation), max_nets_for_sample])
+                if self.sample_size < max_nets_for_sample:
+                    self.sample_size *= (1 + sample_pct_growth)
+                    self.sample_size = np.min(
+                        [int(self.sample_size), max_nets_for_sample])
+                    print(f"\t Sample size increased at step {i+1} to {self.sample_size}")
+                
+                    if convergence_criterion == "hotelling":
+                        hotelling_critical_value = f.ppf(1-hotelling_confidence, num_of_features, self.sample_size - num_of_features) # F(p, n-p) TODO - doc this better
 
                 if sliding_grad_window_k < max_sliding_window_size:
                     sliding_grad_window_k *= (1 + sample_pct_growth)
                     sliding_grad_window_k = np.min(
                         [np.ceil(sliding_grad_window_k).astype(int), max_sliding_window_size])
 
-            grad, hessian = nll_grad_hessian(self._thetas)
+            networks_for_sample = self.generate_networks_for_sample()
+            features_of_net_samples = self._calc_sample_statistics(networks_for_sample)
+            observed_features = self._network_statistics.calculate_statistics(observed_network)
+
+            grad, hessian = self._calculate_optimization_step(observed_features, features_of_net_samples, optimization_method)
+
             if optimization_method == "newton_raphson":
-                try:
-                    inv_hessian = np.linalg.inv(hessian)
-                except np.linalg.LinAlgError:
-                    print("The hessian is not invertible")
-                    inv_hessian = np.linalg.pinv(hessian)
+                inv_hessian = np.linalg.pinv(hessian)
+                # try:
+                #     inv_hessian = np.linalg.inv(hessian)
```

Comment:
We can remove the `inv` attempt (or run `pinv` for matrices with low enough determinant, even if it's not 0)